packages/api/src/router/target-metadata-group.ts

```diff
@@ -24,35 +27,84 @@ export const targetMetadataGroupRouter = createTRPCRouter({
     .meta({
       authorizationCheck: ({ canUser, input }) =>
         canUser
-          .perform(Permission.TargetList)
+          .perform(Permission.TargetMetadataGroupList)
           .on({ type: "workspace", id: input }),
     })
     .input(z.string().uuid())
-    .query(({ ctx, input }) => {
-      return ctx.db
+    .query(async ({ ctx, input }) => {
+      const matchingTargetsQuery = ctx.db
+        .select({ targetId: target.id })
+        .from(target)
+        .leftJoin(targetMetadata, eq(targetMetadata.targetId, target.id))
+        .where(
+          and(
+            eq(target.workspaceId, targetMetadataGroup.workspaceId),
+            sql`${targetMetadata.key} = ANY(${targetMetadataGroup.keys})`,
+          ),
+        )
+        .groupBy(target.id)
+        .having(
+          sql`COUNT(DISTINCT ${targetMetadata.key}) = ARRAY_LENGTH(${targetMetadataGroup.keys}, 1)`,
+        );
+
+      const nonNullGroups = await ctx.db
         .select({
-          targets: sql<number>`count(distinct ${target.id})`.mapWith(Number),
+          targets: sql<number>`
+            COALESCE((
+              SELECT ${count()}
+              FROM (${matchingTargetsQuery}) AS matching_targets
+            ), 0)`.mapWith(Number),
           targetMetadataGroup,
         })
         .from(targetMetadataGroup)
-        .innerJoin(workspace, eq(targetMetadataGroup.workspaceId, workspace.id))
-        .leftJoin(target, eq(target.workspaceId, workspace.id))
-        .leftJoin(targetMetadata, eq(targetMetadata.targetId, target.id))
         .where(
           and(
-            eq(workspace.id, input),
-            sql`"target_metadata"."key" = ANY (${targetMetadataGroup.keys})`,
+            eq(targetMetadataGroup.workspaceId, input),
+            eq(targetMetadataGroup.includeNullCombinations, false),
           ),
         )
-        .groupBy(targetMetadataGroup.id)
         .orderBy(asc(targetMetadataGroup.name));
+
+      const allTargetCount = await ctx.db
```

Comment:
Here I separated the two queries
- first grabbed target groups where null combinations were not allowed and did the metadata key count subquery to get the number of targets in that lable group
- then just counted all the targets in the workspace once, then grabbed _only_ the metadata group where null combinations were allowed. since if null combinations are allowed, then every target would be a match on the group